File: src/main/java/com/freya02/botcommands/annotations/api/prefixed/annotations/JDATextCommand.java
package com.freya02.botcommands.annotations.api.prefixed.annotations;

import com.freya02.botcommands.annotations.api.annotations.BotPermissions;
import com.freya02.botcommands.annotations.api.annotations.Cooldown;
import com.freya02.botcommands.annotations.api.annotations.UserPermissions;
import com.freya02.botcommands.api.prefixed.CommandEvent;
import com.freya02.botcommands.internal.annotations.LowercaseDiscordNamePattern;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Required annotation for bot commands, see all possible options
 * <p>First parameter may be {@link CommandEvent} only for fallback commands</p>
 * <p>Supported parameters needs to be annotated with {@link TextOption @TextOption}</p>
 *
 * @see TextOption
 * @see Hidden
 * @see ID
 * @see BotPermissions
 * @see UserPermissions
 * @see Cooldown
 */
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.METHOD})
public @interface JDATextCommand {
	/**
	 * Primary name of the command, <b>must not contain any spaces</b>
	 *
	 * @return Name of this command
	 */
	@LowercaseDiscordNamePattern
	String name();

	/**
	 * Group name of the command, <b>must not contain any spaces</b>
	 *
	 * @return Group name of this command
	 */
	@LowercaseDiscordNamePattern
	String group() default "";

	/**
	 * Subcommand name of the command, <b>must not contain any spaces</b>
	 *
	 * @return Subcommand name of this command
	 */
	@LowercaseDiscordNamePattern
	String subcommand() default "";

	/**
	 * Specifies the specific order the executable has to be loaded in (1 is the most important)
	 *
	 * @return The order of the method to be loaded in
	 */
	int order() default 0;

	/**
	 * Secondary <b>paths</b> of the command, <b>must not contain any spaces</b>, must follow the same format as slash commands such as <code>name/group/subcommand</code>
	 *
	 * @return Secondary paths of the command
	 */
	String[] aliases() default {};

	/**
	 * Short description of the command, it is displayed in the help command
	 *
	 * @return Short description of the command
	 */
	String description() default "";
}

File: src/main/kotlin/io/github/freya022/botcommands/api/commands/text/annotations/JDATextCommand.kt
package io.github.freya022.botcommands.api.commands.text.annotations

import io.github.freya022.botcommands.api.commands.annotations.*
import io.github.freya022.botcommands.api.commands.text.*
import io.github.freya022.botcommands.api.commands.text.builder.TextCommandBuilder
import io.github.freya022.botcommands.api.commands.text.builder.TextCommandVariationBuilder
import io.github.freya022.botcommands.api.localization.annotations.LocalizationBundle
import io.github.freya022.botcommands.api.localization.context.TextLocalizationContext
import io.github.freya022.botcommands.api.parameters.ParameterResolver
import io.github.freya022.botcommands.api.parameters.resolvers.ICustomResolver
import io.github.freya022.botcommands.api.parameters.resolvers.TextParameterResolver
import net.dv8tion.jda.internal.utils.Checks

/**
 * Declares this function as a text command.
 *
 * ### Text command variations
 * A given text command path (such as `ban temp`) is composed of at least one variation;
 * Each variation has different parameters, and will display separately in the built-in help content.
 *
 * Each variation runs based off its [priority][order],
 * the first variation that has a full match against the user input gets executed.
 *
 * If no regex-based variation (using a [BaseCommandEvent]) matches,
 * the fallback variation is executed (if a variation using [CommandEvent] exists).
 *
 * If no variation matches and there is no fallback,
 * then the [help content][IHelpCommand.onInvalidCommand] is invoked for the command.
 *
 * ### Requirements
 * - The declaring class must be annotated with [@Command][Command] and extend [TextCommand]
 * - First parameter must be [BaseCommandEvent], or, [CommandEvent] for fallback commands/manual token consumption.
 *
 * ### Option types
 * - Input options: Uses [@TextOption][TextOption], supported types are in [ParameterResolver],
 * additional types can be added by implementing [TextParameterResolver].
 * - [TextLocalizationContext]: Uses [@LocalizationBundle][LocalizationBundle].
 * - Custom options and services: No annotation, additional types can be added by implementing [ICustomResolver].
 *
 * @see Command @Command
 * @see Category @Category
 * @see TextOption @TextOption
 * @see Hidden @Hidden
 * @see NSFW @NSFW
 * @see BotPermissions @BotPermissions
 * @see UserPermissions @UserPermissions
 * @see Cooldown @Cooldown
 * @see RateLimit @RateLimit
 * @see Filter @Filter
 *
 * @see TextDeclaration Declaring text commands using the DSL
 *
 * @see TextCommandManager.textCommand DSL equivalent
 */
@Target(AnnotationTarget.FUNCTION)
@Retention(AnnotationRetention.RUNTIME)
annotation class JDATextCommand(
    /**
     * Path components of the command,
     * limited to three components and composed of [`a-zA-Z1-9_-`][Checks.ALPHANUMERIC_WITH_DASH]
     */
    val path: Array<out String>,

    /**
     * Specifies the priority of this text command variation (1 is the most important).
     *
     * By default, if two variations have no order set, parameters are compared between each variation,
     * if one of them is a [String] but the other is not, the [String] parameter is prioritized.
     *
     * If you are using Kotlin, [DSL-declared commands][TextDeclaration] retain the order they are declared in.
     */
    val order: Int = 0,

    /**
     * Secondary **paths** of the command, **must not contain any spaces**,
     * and must follow the same format as slash commands such as `name group subcommand`
     *
     * @see TextCommandBuilder.aliases DSL equivalent
     */
    val aliases: Array<String> = [],

    /**
     * Short description of the command, displayed in the description of the built-in help command.
     *
     * This description can only be set once for a given command path.
     *
     * @see TextCommandBuilder.description DSL equivalent
     */
    val generalDescription: String = "",

    /**
     * Short description of the command displayed in the built-in help command,
     * below the command usage.
     *
     * @see TextCommandVariationBuilder.description DSL equivalent
     */
    val description: String = "",

    /**
     * Usage string for this command variation,
     * the built-in help command already sets the prefix and command name, with a space at the end.
     *
     * If not set, the built-in help command will generate a string out of the options.
     *
     * @see TextCommandVariationBuilder.usage DSL equivalent
     */
    val usage: String = "",

    /**
     * Example command for this command variation,
     * the built-in help command already sets the prefix and command name, with a space at the end.
     *
     * If not set, the built-in help command will generate a string out of the options.
     *
     * @see TextCommandVariationBuilder.example DSL equivalent
     */
    val example: String = ""
)
